engine/src/flutter/shell/platform/darwin/macos/framework/Source/FlutterRunLoop.swift

```diff
@@ -0,0 +1,159 @@
+import Foundation
+
+/// Interface for scheduling tasks on the run loop.
+///
+/// The main difference between using `FlutterRunLoop` to schedule tasks compared
+/// to `DispatchQueue.async` or `RunLoop.perform(_:)` is that `FlutterRunLoop`
+/// schedules the task in both common run loop mode and a private run loop mode,
+/// which allows it to run in a mode where it only processes Flutter messages
+/// (`pollFlutterMessagesOnce()`).
+@objc public class FlutterRunLoop: NSObject {
+  private static let flutterRunLoopMode = CFRunLoopMode("FlutterRunLoopMode" as CFString)
+  private static var _mainRunLoop: FlutterRunLoop?
+
+  private let runLoop: CFRunLoop = CFRunLoopGetCurrent()
+  private var tasks: [Task] = []
+  private let tasksLock = NSLock()
+  private var source: CFRunLoopSource!
+  private var timer: CFRunLoopTimer!
```

Comment:
These two are guaranteed to be non-nil due to the `guard` on creation hence declaring them as implicit unwrapped optionals. Because the create functions take `self` as a parameter, we need to have invoked `super.init()` prior to invoking the create functions, which means these can't be `let` assignments, since they happen after the `super.init()`.

If you've got thoughts on a nicer approach, I'd love to hear it.

Reply:
This is standard if you have to reference to self in the closure - another example is storyboard outlets, as since they are only setup after nibs are hooked up

Reply:
Correct, yep! But this is a class that we generate programmatically, so there's no issues with having to wait for `viewDidLoad` etc. 

I'm thinking more if there's some way we could restructure creation of these -- e.g. create a wrapper class to hold these guys that we could assign (uninitialised) in a let, put `performExpiredTasks` on that class, but the public interface on this (the wrapper) class, and fill the source/timer on that class from our initialiser... but I think that may still have the exact same problem just one level up.

Maybe I'm just averse to this because I'm not super used to it.

Reply:
I think what you are doing is fine, since we don't own the class. 

On the other hand, if we own the class, there are 2 options: 
1. use force unwrap and set after `self.init`, like what you do here
2. redesign our own class such that the closure itself is optional, and we set the closure property separately after its init.